iree/compiler/Dialect/Flow/Transforms/ConvertLinalgMatmulToMmt4D.cpp

```diff
@@ -215,18 +222,56 @@ static bool haveEqualShapeDim(Value x, Value y, int i) {
          y.getType().cast<ShapedType>().getDimSize(i);
 }
 
+// Current instances of integer matmuls in linalg tend to use signless integers
+// to mean signed. Accept both, while this is sorted out.
+static bool isSignedOrSignlessInteger(Type type, int width) {
+  return type.isSignedInteger(width) || type.isSignlessInteger(width);
+}
+
+class Mmt4DTileParams {
+public:
+  Mmt4DTileParams(int64_t M0, int64_t K0, int64_t N0, std::string comment) : M0(M0), K0(K0), N0(N0), comment(comment) {}
+  std::array<int64_t, 2> lhs() const { return {M0, K0}; }
+  std::array<int64_t, 2> rhs() const { return {K0, N0}; }
+  std::array<int64_t, 2> acc() const { return {M0, N0}; }
+  const std::string& getComment() const { return comment; }
+private:
+  const int64_t M0;
+  const int64_t K0;
+  const int64_t N0;
+  const std::string comment;
+};
+
+static Mmt4DTileParams chooseTileParamsForElemTypesAndTargetInfo(Type lhsElemType, Type rhsElemType, Type accElemType, const ConvertMatmulToMmt4DTargetInfo& target_info)
+{
+  if (isSignedOrSignlessInteger(lhsElemType, 8) &&
+      isSignedOrSignlessInteger(rhsElemType, 8) &&
+      isSignedOrSignlessInteger(accElemType, 32) &&
+      target_info.aarch64 &&
+      target_info.dotprod) {
+    return Mmt4DTileParams(8, 4, 8, "i8*i8->i32, aarch64, dotprod");
+  }
+  return Mmt4DTileParams(8, 2, 4, "generic tiling parameters, as no known kernel was matched for this matmul and target");
```

Comment:
Do we want to do generic tiling, or do we want to bail and not mmt4d-ify those? (actual question, I'm not sure)

Reply:
it's a good question and i've been flipflopping on this, but, thanks for raising it again because now i think i know where I want to go.

the reason to do generic tiling is: test coverage. For example, at the moment, our CI runs on x86 where we dont have any kernel, so generic tiling is all we do. Drop this, drop all CI coverage for mmt4d.

but from the perspective of real users, it's much better to not do generic ( = slow) tiling and only allow mmt4d to kick in where it's going to be really optimized. this would unlock enabling this pattern by default, trusting that it's either beneficial or a no-op.

so i think we need another boolean pass option here, for generic tiling. we would enable it in tests, but it would otherwise be off by default.

Reply:
Implemented (`enable_generic_slow` boolean)